pageserver/src/tenant/delete.rs

```diff
@@ -538,16 +541,71 @@ impl DeleteTenantFlow {
+                // Under the TenantMap lock, try to remove the tenant.  We usually succeed, but if
+                // we encounter an InProgress marker, yield the barrier it contains and wait on it.
+                let barrier = {
+                    let mut locked = tenants.write().unwrap();
+                    let removed = locked.remove(&tenant.tenant_id);
+
+                    // FIXME: we should not be modifying this from outside of mgr.rs.
+                    // This will go away when we simplify deletion (https://github.com/neondatabase/neon/issues/5080)
+                    crate::metrics::TENANT_MANAGER
+                        .tenant_slots
+                        .set(locked.len() as u64);
+
+                    match removed {
+                        TenantsMapRemoveResult::Occupied(slot) => {
+                            match slot {
+                                TenantSlot::Secondary => {
+                                    // This is unexpected: this secondary tenants should not have been created, and we
+                                    // are not in a position to shut it down from here.
+                                    tracing::warn!(
+                                        "Tenant transitioned to secondary mode while deleting!"
+                                    );
+                                }
+                                TenantSlot::Attached(tenant) => match tenant.current_state() {
+                                    TenantState::Stopping { .. } | TenantState::Broken { .. } => {
+                                        // Expected: we put the tenant into stopping state before we start deleting it
+                                    }
+                                    state => {
+                                        // Unexpected state
+                                        tracing::warn!(
+                                            "Tenant in unexpected state {state} after deletion"
+                                        );
+                                    }
+                                },
+                                TenantSlot::InProgress(_) => {
+                                    unreachable!("TenantsMap::remove handles InProgress separately, should never return it here");
+                                }
+                            };
+                            break;
+                        }
+                        TenantsMapRemoveResult::Vacant => {
+                            tracing::warn!(
+                                "Tenant removed from TenantsMap before deletion completed"
+                            );
+                            break;
+                        }
+                        TenantsMapRemoveResult::InProgress(barrier) => {
+                            // An InProgress entry was found, we must wait on its barrier
+                            barrier
+                        }
+                    }
+                };
```

Comment:
Code readability wise, I think structuring it as:
```match removed {
            TenantsMapRemoveResult::Occupied(TenantSlot::Attached(tenant)) => {
                if let TenantState::Stopping { .. } | TenantState::Broken { .. } = tenant.current_state() {
                } else {
                }
                break;
            }
            TenantsMapRemoveResult::Occupied(TenantSlot::Secondary) => {
                break;
            }
            TenantsMapRemoveResult::Vacant => {
                break;
            }
            TenantsMapRemoveResult::InProgress(barrier) => {
                barrier
            }
            _ => unreachable!("TenantsMap::remove handles InProgress separately, should never return it here"),
        }```
 would be easier to follow

Reply:
Flattened this in 476f869d88850240e8d396bcf2631a467a011fbb